whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,213 @@
+import net.ripe.db.whois.common.iptree.Ipv6Tree;
+import net.ripe.db.whois.common.rpsl.AttributeType;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+import net.ripe.db.whois.common.rpsl.RpslObject;
+import net.ripe.db.whois.common.rpsl.attrs.Domain;
+import net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus;
+import net.ripe.db.whois.common.rpsl.attrs.InetnumStatus;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static net.ripe.db.whois.common.rpsl.attrs.Inet6numStatus.ALLOCATED_BY_RIR;
+import static net.ripe.db.whois.common.rpsl.attrs.InetnumStatus.ALLOCATED_UNSPECIFIED;
+
+
+@Service
+public class RdapRelationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RdapRelationService.class);
+
+    private final Ipv4Tree ip4Tree;
+    private final Ipv6Tree ip6Tree;
+    private final Ipv4DomainTree ipv4DomainTree;
+    private final Ipv6DomainTree ipv6DomainTree;
+    private final RpslObjectDao rpslObjectDao;
+
+    @Autowired
+    public RdapRelationService(final Ipv4Tree ip4Tree, final Ipv6Tree ip6Tree,
+                               final Ipv4DomainTree ipv4DomainTree, final Ipv6DomainTree ipv6DomainTree,
+                               final RpslObjectDao rpslObjectDao) {
+        this.ip4Tree = ip4Tree;
+        this.ip6Tree = ip6Tree;
+        this.ipv4DomainTree = ipv4DomainTree;
+        this.ipv6DomainTree = ipv6DomainTree;
+        this.rpslObjectDao = rpslObjectDao;
+    }
+
+    public List<String> getDomainRelationPkeys(final String pkey, final RelationType relationType){
+        final Domain domain = Domain.parse(pkey);
+        final IpInterval reverseIp = domain.getReverseIp();
+        final List<IpEntry> ipEntries = getIpEntries(getIpDomainTree(reverseIp), relationType, reverseIp);
+
+        return ipEntries
+                .stream()
+                .map(ipEntry -> rpslObjectDao.getById(ipEntry.getObjectId()).getKey().toString())
+                .toList();
+    }
+
+    public List<String> getInetnumRelationPkeys(final String pkey, final RelationType relationType){
+        final IpInterval ip = IpInterval.parse(pkey);
+        final List<IpEntry> ipEntries = getIpEntries(getIpTree(ip), relationType, ip);
+        return ipEntries
+                .stream()
+                .map(ipEntry -> ipEntry.getKey().toString())
+                .toList();
+    }
+
+    private List<IpEntry> getIpEntries(final IpTree ipTree, final RelationType relationType,
+                                       final IpInterval reverseIp) {
+        return switch (relationType) {
+            case UP -> List.of(searchFirstLessSpecific(ipTree, reverseIp));
+            case TOP -> searchTopLevelResource(ipTree, reverseIp);
```

Comment:
TOP should be single resource like UP, not a list? 
" If the server receives a search containing the relation value "top",
   it will return the top object for the specified INR object value as
   though that object had been requested directly. "

Reply:
Also TOP should be getLast from findAllLessSpecific?

Reply:
our top level is always administrative resource
for IPv4 is 0.0.0.0 - 255.255.255.255 always
for IPv6 is ::/0 always

For IPv4 is easy, we get the top level removing that one which has "ALLOCATED UNSPECIFIED" status
For IPv6 to detect the administrative resource is not that easy. An administrative resource is the one with "ALLOCATED-BY-RIR" status when the children is "ALLOCATED-BY-RIR" as well.

If we hit a resource with "ALLOCATED-BY-RIR" we cannot directly find the firstMoreSpecific, because they can be more than 1.
So we need to iterate and search for UP. If the Xresource is "ALLOCATED-BY-RIR" and his parent is "ALLOCATED-BY-RIR" as well...then the parent is an administrative resource and we return the Xresource.